x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingService.java

```diff
@@ -0,0 +1,251 @@
+import org.elasticsearch.xpack.ml.inference.persistence.TrainedModelProvider;
+
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ModelLoadingService implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(ModelLoadingService.class);
+    // TODO should these be ConcurrentHashMaps if all interactions are synchronized?
+    private final ConcurrentHashMap<String, Optional<Model>> loadedModels = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, Queue<ActionListener<Model>>> loadingListeners = new ConcurrentHashMap<>();
+    private final TrainedModelProvider provider;
+    private final ThreadPool threadPool;
+
+    public ModelLoadingService(TrainedModelProvider trainedModelProvider, ThreadPool threadPool, ClusterService clusterService) {
+        this.provider = trainedModelProvider;
+        this.threadPool = threadPool;
+        clusterService.addListener(this);
+    }
+
+    public void getModel(String modelId, long modelVersion, ActionListener<Model> modelActionListener) {
+        String key = modelKey(modelId, modelVersion);
+        Optional<Model> cachedModel = loadedModels.get(key);
+        if (cachedModel != null) {
+            if (cachedModel.isPresent()) {
+                modelActionListener.onResponse(cachedModel.get());
+                return;
+            }
+        }
+        if (loadModelIfNecessary(key, modelId, modelVersion, modelActionListener) == false) {
+            // If we the model is not loaded and we did not kick off a new loading attempt, this means that we may be getting called
+            // by a simulated pipeline
+            logger.debug("[{}] version [{}] not actively loading, eager loading without cache", modelId, modelVersion);
+            provider.getTrainedModel(modelId, modelVersion, ActionListener.wrap(
+                trainedModelConfig ->
+                    modelActionListener.onResponse(new LocalModel(trainedModelConfig.getModelId(), trainedModelConfig.getDefinition())),
+                modelActionListener::onFailure
+            ));
+        } else {
+            logger.debug("[{}] version [{}] is currently loading, added new listener to queue", modelId, modelVersion);
+        }
+    }
+
+    /**
+     * Returns true if the model is loaded and the listener has been given the cached model
+     * Returns true if the model is CURRENTLY being loaded and the listener was added to be notified when it is loaded
+     * Returns false if the model is not loaded or actively being loaded
+     */
+    private boolean loadModelIfNecessary(String key, String modelId, long modelVersion, ActionListener<Model> modelActionListener) {
+        synchronized (loadingListeners) {
```

Comment:
I am not sure this means that we are loading one model at a time. This method does the following while in the `synchronized` block

* Check if we have successfully loaded the model and get it.
* If there is a failed load attempt, create a new queue of listeners (adding the new listener to the queue) and kick of an asynchronous loading of the model and exit.
* If we have not attempted to load the model (failed or success), see if there are existing listeners (indicates a loading attempt is in progress) and add the new listener and exit.

Since `load_model` is an asynchronous method, we exit the `synchronized` block here and then enter another `synchronized` block later in `handleLoadSuccess` or `handleLoadFailure`.